src/mito2/src/sst/index.rs

```diff
@@ -384,11 +399,221 @@ impl IndexerBuilderImpl {
+        }
+    }
+}
+
+/// Outcome of an index build task.
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub enum IndexBuildOutcome {
+    Finished,
+    Aborted(String),
+}
+
+pub struct IndexBuildTask {
+    /// The file meta to build index for.
+    pub file_meta: FileMeta,
+    pub reason: IndexBuildType,
+    pub flushed_entry_id: Option<EntryId>,
+    pub flushed_sequence: Option<SequenceNumber>,
+    pub committed_sequence: Option<SequenceNumber>,
+    pub access_layer: AccessLayerRef,
+    pub write_cache: Option<WriteCacheRef>,
+    pub file_purger: FilePurgerRef,
+    /// When write cache is enabled, the indexer builder should be built from the write cache.
+    /// Otherwise, it should be built from the access layer.
+    pub indexer_builder: Arc<dyn IndexerBuilder + Send + Sync>,
+    /// Request sender to notify the region worker.
+    pub(crate) request_sender: mpsc::Sender<WorkerRequestWithTime>,
+    /// Optional sender to send the result back to the caller.
+    pub result_sender: Option<oneshot::Sender<IndexBuildOutcome>>,
+}
+
+impl IndexBuildTask {
+    fn into_index_build_job(mut self) -> Job {
+        Box::pin(async move {
+            self.do_index_build().await;
+        })
+    }
+
+    async fn do_index_build(&mut self) {
+        let outcome = match self.index_build().await {
+            Ok(outcome) => outcome,
+            Err(e) => {
+                warn!(
+                    e; "Index build task failed, region: {}, file_id: {}",
+                    self.file_meta.region_id, self.file_meta.file_id,
+                );
+                IndexBuildOutcome::Aborted(format!("Index build failed: {}", e))
+            }
+        };
+        if let Some(sender) = self.result_sender.take() {
+            let _ = sender.send(outcome);
+        }
+    }
+
+    async fn check_sst_file_exists(&self) -> bool {
+        let sst_path = location::sst_file_path(
+            self.access_layer.table_dir(),
+            RegionFileId::new(self.file_meta.region_id, self.file_meta.file_id),
+            self.access_layer.path_type(),
+        );
+        match self.access_layer.object_store().exists(&sst_path).await {
+            Ok(exists) => {
+                if !exists {
+                    warn!(
+                        "SST file not found during index build, skipping. region: {}, file_id: {}",
+                        self.file_meta.region_id, self.file_meta.file_id
+                    );
+                }
+                exists
+            }
+            Err(e) => {
+                warn!(
+                    "Failed to check SST file existence during index build, skipping. region: {}, file_id: {}, error: {:?}",
+                    self.file_meta.region_id, self.file_meta.file_id, e
+                );
+                false
+            }
+        }
+    }
+
+    async fn index_build(&mut self) -> Result<IndexBuildOutcome> {
+        let mut indexer = self.indexer_builder.build(self.file_meta.file_id).await;
```

Comment:
We assume puffin/parquet files in the engine are immutable. It'd be better to assign a new id to the index if the file already has an index. We can consider it after this PR.